Add Swidget integration

michaelkkehoe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
No breaking changes
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
This change introduces a new integration/ platform called 'Swidget' (https://www.swidget.com/collections/all-products). Swidget has plug and play inserts that can be inserted into host modules. This commit adds the integration and creates support for the 'light' entity. There will be additional changes coming for sensor/ binary_sensor/ button/ switch.
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes # N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/32324

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,54 @@
+    hass.data.setdefault(DOMAIN, {})
+    try:
+        device = await discover_single(
+            host=entry.data["host"],
```
please use `CONF_HOST` and `CONF_PASSWORD`

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,54 @@
+    except SwidgetException as ex:
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN][entry.entry_id] = SwidgetDataUpdateCoordinator(hass, device)
```
do a `await coordinator.async_config_entry_first_refresh()` before adding it to the entry

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+DISCOVERY_INTERVAL = timedelta(minutes=15)
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_HOST, default=""): str,
```
```suggestion
        vol.Optional(CONF_HOST): str,
```

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+class SwidgetConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for swidget."""
+
+    VERSION = 1
```
```suggestion
```
is default

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
```
are you planning to re-use validate input? If not, just do it inline, can also save you the extra exceptions

michaelkkehoe replied:
I am. There are SSDP/ DHCP discovery mechanisms that will be added in the future that will reuse the validate_input() functionality

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+            use_https=True,
+            use_websockets=False,
+        )
+        await device.update()
```
can this not throw errors?

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+        await device.update()
+        friendly_name: str = device.friendly_name
+        unique_id: str = f"{device.id}_{device.mac_address}"
+        await device.stop()
```
idem

zweckj replied:
disregard

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,54 @@
+
+    hass.data[DOMAIN][entry.entry_id] = SwidgetDataUpdateCoordinator(hass, device)
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    await device.start()
```
idem

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,54 @@
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    await device.start()
+    hass.loop.create_task(device.get_websocket().listen())
+    await device.update()
```
idem

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/diagnostics.py:
```diff
@@ -0,0 +1,19 @@
+"""Diagnostics support for Swidet."""
```
add those later in a separate PR

michaelkkehoe replied:
Removed this.

zweckj replied:
still here 😉

zweckj commented on homeassistant/components/swidget/entity.py:
```diff
@@ -0,0 +1,57 @@
+from .const import DOMAIN
+from .coordinator import SwidgetDataUpdateCoordinator
+
+_T = TypeVar("_T", bound="CoordinatedSwidgetEntity")
+_P = ParamSpec("_P")
```
why the generics?

michaelkkehoe replied:
Removed

zweckj commented on homeassistant/components/swidget/light.py:
```diff
@@ -0,0 +1,96 @@
+
+    device: SwidgetDimmer
+
+    @async_refresh_after
```
firstly, I'm not sure I like the decorator just to avoid a single `self.async_request_refresh` call, because I'd argue that call is clearer than the decorator. Secondly, do you really need to do an update after each call? Can't you assume that value was set if the call didn't throw errors?

michaelkkehoe replied:
Removed.

zweckj commented on homeassistant/components/swidget/light.py:
```diff
@@ -0,0 +1,96 @@
+        return modes
+
+    @property
+    def color_mode(self) -> ColorMode:
```
use the `attr_` here instead

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/light.py:
```diff
@@ -0,0 +1,96 @@
+        await self.device.turn_on()
+
+    @property
+    def supported_color_modes(self) -> set[ColorMode] | set[str] | None:
```
I think we can also use the `attr_` for brevity

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+
+    async def _async_update_data(self) -> dict[str, Any]:
+        """Fetch all device and sensor data from api."""
+        # No need to call any explicit update function here. The device will update the state itself
```
but you are updating something here? If you really don't need to can't we just stop the polling?

zweckj commented on homeassistant/components/swidget/manifest.json:
```diff
@@ -0,0 +1,26 @@
+  ],
+  "documentation": "https://www.home-assistant.io/integrations/swidget",
+  "integration_type": "device",
+  "iot_class": "local_push",
```
where is the push happening?

michaelkkehoe replied:
The Swidget API client creates a websocket to the device. The device will automatically push change notifications back to the Swidget API client over websocket (so you don't have to poll it for changes).

zweckj commented on homeassistant/components/swidget/entity.py:
```diff
@@ -0,0 +1,57 @@
+
+class CoordinatedSwidgetEntity(CoordinatorEntity[SwidgetDataUpdateCoordinator]):
+    """Common base class for all coordinated entities."""
+
```
```suggestion
   _attr_has_entity_name=True
```
needs to be set for new integrations

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/light.py:
```diff
@@ -0,0 +1,96 @@
+) -> None:
+    """Set up light."""
+    coordinator: SwidgetDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    if coordinator.device.is_dimmer:
```
what if the device is not a dimmer? It doesn't have any entities? Why even add it at all

michaelkkehoe replied:
Swidget (currently) have three types of devices: dimmer, outlet, on/off switch. We only set up the light entity if the device is a dimmer

zweckj commented on tests/components/swidget/test_config_flow.py:
```diff
@@ -0,0 +1,63 @@
+        )  # Ensure title is correctly returned
+
+        # Test validation failure (cannot connect)
+        with patch(
```
please fix the exception and continue the test (if in a non-terminal state) to show we can recover from errors. I'd suggest you switch to parametrized tests for the exceptions

michaelkkehoe replied:
This should be fixed.

zweckj commented on tests/components/swidget/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Setup for tests used for testing the Swidget integration."""
```
While (afaik) tests for your platforms are no hard requirement atm, I would suggest you add test cases for them.

michaelkkehoe replied:
There will be a significant test suite added in the near future.

zweckj reviewed (changes requested):
Thanks for your contribution 👍First feedback

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,54 @@
+    hass.data[DOMAIN][entry.entry_id] = SwidgetDataUpdateCoordinator(hass, device)
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    await device.start()
+    hass.loop.create_task(device.get_websocket().listen())
```
```suggestion
    hass.async_create_background_task((device.get_websocket().listen(), "websocket_connection")
```
I think this method is preferred, instead of directly accessing the event loop

michaelkkehoe replied:
Thanks

autinerd committed:
Add `__pycache__` to gitignore (#114056)

joostlek committed:
Refactor Aurora tests (#117323)

epenet committed:
Use remove_device helper in tasmota tests (#116617)

scop committed:
Filter out HTML greater/less than entities from huawei_lte sensor values (#117209)

epenet committed:
Allow manual delete of stale Renault vehicles (#116229)

mib1185 committed:
Consider only active config entries as media source in Synology DSM (#117691)

consider only active config entries as media source

mib1185 committed:
Move entity definitions into own module in AVM Fritz!Tools (#117701)

* move entity definitions into own module

* merge entity description mixin

* add entity.py to .coveragerc

balloob committed:
Add GitHub CoPilot to extensions devcontainer (#117699)

balloob committed:
Remove AngellusMortis as code-owner Unifi Protect (#117708)

thecode committed:
Avoid Shelly RPC reconnect during device shutdown (#117702)

balloob committed:
Add API class to LLM helper (#117707)

* Add API class to LLM helper

* Add more tests

* Rename intent to assist to broaden scope

bieniu committed:
Use SnmpEngine stored in hass.data by singleton in Brother integration (#117043)

jjlawren committed:
Bump PlexAPI to 4.15.13 (#117712)

JakeMartin-ICL committed:
Bump monzopy to 1.2.0 (#117730)

Shutgun committed:
Use reauth helper in devolo Home Network (#117736)

thecode committed:
Bump aioshelly to 10.0.0 (#117728)

joostlek committed:
Use helper in Withings reauth (#117727)

Shutgun committed:
Use reauth helper in devolo Home Control (#117739)

Anrijs committed:
Bump aranet4 to 2.3.4 (#117738)

bump aranet4 lib version to 2.3.4

bieniu committed:
Refactor tests for Brother integration (#117377)

* Refactor tests - step 1

* Remove fixture

* Refactor test_init

* Refactor test_diagnostics

* Refactor test_config_flow

* Increase test coverage

* Cleaning

* Cleaning

* Check config entry state in test_async_setup_entry

* Simplify patching

* Use AsyncMock when patching

---------

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

jpbede committed:
Bump pydiscovergy to 3.0.1 (#117740)

bdraco committed:
Fix setting MQTT socket buffer size with WebsocketWrapper (#117672)

Shulyaka committed:
LLM Tools support for Google Generative AI integration (#117644)

* initial commit

* Undo prompt chenges

* Move format_tool out of the class

* Only catch HomeAssistantError and vol.Invalid

* Add config flow option

* Fix type

* Add translation

* Allow changing API access from options flow

* Allow model picking

* Remove allowing HASS Access in main flow

* Move model to the top in options flow

* Make prompt conditional based on API access

* convert only once to dict

* Reduce debug logging

* Update title

* re-order models

* Address comments

* Move things

* Update labels

* Add tool call tests

* coverage

* Use LLM APIs

* Fixes

* Address comments

* Reinstate the title to not break entity name

---------

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

albertogeniola committed:
Update elmax_api to v0.0.5 (#117693)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

RicArch97 committed:
Bump crownstone-sse to 2.0.5, crownstone-cloud to 1.4.11 (#117748)

epenet committed:
Add turn_on to SamsungTV remote (#117403)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Fix missing type for mqtt websocket wrapper (#117752)

bdraco committed:
Block older versions of custom integration mydolphin_plus since they cause crashes (#117751)

bdraco committed:
Fix race in config entry setup (#117756)

bdraco committed:
Add setup time detail to diagnostics (#117766)

cdce8p committed:
Update pylint to 3.2.2 (#117770)

epenet committed:
Move environment_canada coordinator to separate module (#117426)

cdce8p committed:
Use PEP 695 for class annotations (1) (#117775)

cdce8p committed:
Use PEP 695 for class annotations (2) (#117776)

cdce8p committed:
Use PEP 695 for class annotations (4) (#117778)

cdce8p committed:
Use PEP 695 for class annotations (3) (#117777)

cdce8p committed:
Use PEP 695 for hass_dict annotations (#117779)

cdce8p committed:
Use PEP 695 for covariant class annotations (#117780)

bdraco committed:
Fix blocking I/O in the event loop when loading timezones (#117721)

epenet committed:
Downgrade point quality scale to silver (#117783)

epenet committed:
Downgrade tellduslive quality scale to silver (#117784)

mawoka-myblock committed:
Set integration_type to device for apsystems integration (#117782)

cdce8p committed:
Update mypy-dev to 1.11.0a3 (#117786)

cdce8p committed:
Use PEP 695 misc (#117788)

cdce8p committed:
Use PEP 695 for function annotations with scoping (#117787)

frenck committed:
Update wled to 0.18.0 (#117790)

cdce8p committed:
Enable UP040 ruff check (#117792)

synesthesiam committed:
Unpause media players that were paused outside voice (#117575)

* Unpause media players that were paused outside voice

* Use time.time()

* Update last paused as media players change state

* Add sleep to test

* Use context

* Implement suggestions

cdce8p committed:
Use PEP 695 type alias in tests (#117797)

cdce8p committed:
Remove quotes surrounding annotations (#117817)

MatthewFlamm committed:
Bump pynws to 1.8.1 for nws (#117820)

bdraco committed:
Detect incorrect exception in forwarded platforms (#117754)

* Detect incorrect exception in forwarded platforms

If an integration raises ConfigEntryError/ConfigEntryAuthFailed/ConfigEntryAuthFailed
in a forwarded platform it would affect the state of the config entry and cause it to
process unloads and setup retries in while the other platforms continued to setup

* Detect incorrect exception in forwarded platforms

If an integration raises ConfigEntryError/ConfigEntryAuthFailed/ConfigEntryAuthFailed
in a forwarded platform it would affect the state of the config entry and cause it to
process unloads and setup retries in while the other platforms continued to setup

* Update homeassistant/config_entries.py

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

* adjust

* fix

---------

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

puddly committed:
Account for disabled ZHA discovery config entries when migrating SkyConnect integration (#117800)

* Properly handle disabled ZHA discovery config entries

* Update tests/components/homeassistant_sky_connect/test_util.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

---------

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

bdraco committed:
Bump dbus-fast to 2.21.3 (#117824)

bdraco committed:
Bump tesla-powerwall to 0.5.2 (#117823)

bdraco committed:
Improve error message when template is rendered from wrong thread (#117822)

* Improve error message when template is rendered from wrong thread

* Improve error message when template is rendered from wrong thread

cdce8p committed:
Update scaffold templates to use runtime_data (#117819)

cdce8p committed:
Improve zwave_js TypeVar usage (#117810)

* Improve zwave_js TypeVar usage

* Use underscore for TypeVar name

allenporter committed:
Stop the nest subscriber on Home Assistant stop (#117830)

dependabot[bot] committed:
Bump codecov/codecov-action from 4.4.0 to 4.4.1 (#117836)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

epenet committed:
Refactor SamsungTV auth check (#117834)

dependabot[bot] committed:
Bump github/codeql-action from 3.25.5 to 3.25.6 (#117835)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bajansen committed:
Tesla wall connector add sensors (#117769)

bdraco committed:
Remove @ from codeowners when downloading diagnostics (#117825)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

cdce8p committed:
Remove future import to fix broken typing.get_type_hints call (#117837)

Thomas55555 committed:
Bump aioautomower to 2024.5.1 (#117815)

bdraco committed:
Convert solax to use DataUpdateCoordinator (#117767)

cdce8p committed:
Use PEP 695 misc (2) (#117814)

edenhaus committed:
Use uv instead of pip in development env (#113517)

bajansen committed:
Tesla Wall Connector fix spelling error/typo (#117841)

epenet committed:
Remove silver integrations from NO_DIAGNOSTICS (#117840)

epenet committed:
Simplify samsungtv unload (#117838)

bdraco committed:
Small speed up to adding event bus listeners (#117849)

bdraco committed:
Simplify service description cache logic (#117846)

bdraco committed:
Cache area registry JSON serialize (#117847)

We already cache the entity and device registry, but since I never
used area until recently I did not have enough to notice that they
were not cached

bdraco committed:
Speed up async_get_loaded_integrations (#117851)

* Speed up async_get_loaded_integrations

Use a setcomp and difference to find the components to split
to avoid the loop. A setcomp is inlined in python3.12 so its
much faster

* Speed up async_get_loaded_integrations

Use a setcomp and difference to find the components to split
to avoid the loop. A setcomp is inlined in python3.12 so its
much faster

* simplify

* fix compat

* bootstrap

* fix tests

bdraco committed:
Small speed up to logbook humanify (#117854)

wittypluck committed:
Add support for Glances v4 (#117664)

synesthesiam committed:
Add description to intent handlers and use in LLM helper (#117864)

mkmer committed:
Add missing placeholder name to reauth (#117869)

add placeholder name to reauth

balloob committed:
Address late feedback Google LLM (#117873)

balloob committed:
Make Google AI model picker a dropdown (#117878)

autinerd committed:
Enable Ruff RET502 (#115139)

farmio committed:
Add `async_turn_on/off` methods for KNX climate entities (#117882)

Add async_turn_on/off methods for KNX climate entities

farmio committed:
Add data point type option to `knx.telegram` trigger (#117463)

* Add data point type (dpt) option to `knx.telegram` trigger

* Rename from `dpt` to `type` to match services

* Add test for GroupValueRead telegrams

* Fix device trigger schema inheritance

* Typesafe dispatcher signal

* readability

* Avoid re-decoding with same transcoder

farmio committed:
Add value_template option to KNX expose (#117732)

* Add value_template option to KNX expose

* template exception handling

elupus committed:
Update philips_js to 3.2.1 (#117881)

* Update philips_js to 3.2.0

* Update to 3.2.1

bdraco committed:
Simplify and speed up mqtt_config_entry_enabled check (#117886)

bdraco committed:
Fix mqtt timer churn (#117885)

Borrows the same design from homeassistant.helpers.storage to avoid
rescheduling the timer every time async_schedule is called if a timer
is already running.

Instead of the timer fires too early it gets rescheduled for the time
we wanted it. This avoids 1000s of timer add/cancel during startup

bdraco committed:
Reduce overhead to call get_mqtt_data (#117887)

We call this 100000s of times if there are many subscriptions

https://github.com/home-assistant/core/pull/109030#issuecomment-2123612530

Shulyaka committed:
LLM Tools: Add device_id (#117884)

bdraco committed:
Avoid creating tasks to subscribe to discovery in MQTT (#117890)

Shulyaka committed:
LLM Tools support for OpenAI integration (#117645)

* initial commit

* Add tests

* Move tests to the correct file

* Fix exception type

* Undo change to default prompt

* Add intent dependency

* Move format_tool out of the class

* Fix tests

* coverage

* Adjust to new API

* Update strings

* Update tests

* Remove unrelated change

* Test referencing non-existing API

* Add test to verify no exception on tool conversion for Assist tools

* Bump voluptuous-openapi==0.0.4

* Add device_id to tool input

* Fix tests

---------

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

bdraco committed:
Reduce overhead to validate mqtt topics (#117891)

* Reduce overhead to validate mqtt topics

valid_topic would iterate all the chars 4x, refactor to only
do it 1x

valid_subscribe_topic would enumerate all the chars when there was
no + in the string

* check if adding a cache helps

* tweak lrus based on testing stats

* note to future maintainers

* note to future maintainers

* keep standard lru_cache size as increasing makes no material difference

Kane610 committed:
Support carbon dioxide and formaldehyde sensors in deCONZ (#117877)

* Add formaldehyde sensor
* Add carbon dioxide sensor
* Bump pydeconz to v116

bdraco committed:
Small speed up to creating flows (#117896)

Use a defaultdict instead of setdefault

jschlyter committed:
Improve typing via hassfest serializer (#117382)

epenet committed:
Add coordinator to SamsungTV (#117863)

* Introduce samsungtv coordinator

* Adjust

* Adjust media_player

* Remove remote

* Adjust

* Fix mypy

* Adjust

* Use coordinator.async_refresh

epenet committed:
Adjust conftest type hints (#117900)

MatthewFlamm committed:
Keep observation data valid for 60 min and retry with no data for nws (#117109)

Co-authored-by: J. Nick Koston <nick@koston.org>

epenet committed:
Add MAC cleanup to SamsungTV (#117906)

* Add MAC cleanup to samsungtv

* Simplify

* Adjust

* leftover

* Appl

Co-authored-by: J. Nick Koston <nick@koston.org>

* Update diagnostics tests

---------

Co-authored-by: J. Nick Koston <nick@koston.org>

tichris0 committed:
Adopt new runtime entry data model for AlarmDecoder (#117856)

* Adopt new runtime entity data model for AlarmDecoder

Transition the AlarmDecoder integration to the new runtime entity model.

* Apply change suggestions by epenet

Tested & applied the suggestions from epenet.

bdraco committed:
Refactor MQTT to replace get_mqtt_data with HassKey (#117899)

bdraco committed:
Bump habluetooth to 3.1.0 (#117905)

epenet committed:
Remove duplicate code in SamsungTV (#117913)

joostlek committed:
Exclude modbus from diagnostics hassfest check (#117855)

cdce8p committed:
Fix performance regression with SignalType (#117920)

gjohansson-ST committed:
Add clear night to smhi (#115998)

mkmer committed:
Use common title for reauth confirm in Whirlpool config flow (#117924)

* Add missing placeholder

* Use common title for reauth

tronikos committed:
Google Generative AI: Handle response with empty parts in generate_content (#117908)

Handle response with empty parts in generate_content

OnFreund committed:
Bump pyrympro to 0.0.8 (#117919)

Thomas55555 committed:
Add switch for stay out zones in Husqvarna Automower (#117809)

Co-authored-by: Robert Resch <robert@resch.dev>

synesthesiam committed:
Fail if targeting all devices in the house in service intent handler (#117930)

* Fail if targeting all devices in the house

* Update homeassistant/helpers/intent.py

---------

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

bdraco committed:
Speed up registry indices (#117897)

* Use defaultdict for registry indices

defaultdict is faster and does not have to create an empty
dict that gets throw away when the key is already present

* Use defaultdict for registry indices

defaultdict is faster and does not have to create an empty
dict that gets throw away when the key is already present

puddly committed:
Simplify SkyConnect setup flow (#117868)

* Delay firmware probing until after the user picks the firmware type

* Remove confirmation step

* Fix unit tests

* Simplify unit test patching logic

Further simplify unit tests

* Bump Zigbee firmware up to the first choice

* Reuse `async_step_pick_firmware` during options flow

* Proactively validate all ZHA entries, not just the first
There can only be one (for now) so this changes nothing functionally

* Add unit test for bad firmware when configuring Thread

joostlek committed:
Add lock to token validity check (#117912)

allenporter committed:
Move nest diagnostic tests to use snapshots (#117929)

bdraco committed:
Migrate issue registry to use singleton helper (#117848)

* Migrate issue registry to use singleton helper

The other registries were already migrated, but since this
one had a read only flag, it required a slightly different
solution since it uses the same hass.data key

* refactor

epenet committed:
Add available and state to SamsungTV remote (#117909)

* Add available and state to SamsungTV remote

* Align turn_off

* Fix merge

* Fix merge (again)

mib1185 committed:
Store runtime data inside the config entry in iBeacon (#117936)

store runtime data inside the config entry

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Add number entities for screenlogic values used in SI calc (#117812)

swoga committed:
Prevent time pattern reschedule if cancelled during job execution (#117879)

Co-authored-by: J. Nick Koston <nick@koston.org>

gnumpi committed:
Send MEDIA_ANNOUNCE flag to ESPHome media_player (#116993)

jbouwh committed:
Refactor shared mqtt schema's to new module (#117944)

* Refactor mqtt schema's to new module

* Remove unrelated change

balloob committed:
Update LLM no tools message (#117935)

bdraco committed:
Refactor state_reported listener setup to avoid merge in async_fire_internal (#117953)

* Refactor state_reported listener setup to avoid merge in async_fire_internal

Instead of merging the listeners in async_fire_internal, setup the listener for
state_changed at the same time so async_fire_internal can avoid having to copy
another list

* Refactor state_reported listener setup to avoid merge in async_fire_internal

Instead of merging the listeners in async_fire_internal, setup the listener for
state_changed at the same time so async_fire_internal can avoid having to copy
another list

* tweak

* tweak

* tweak

* tweak

* tweak

thecode committed:
Add Shelly debug logging for async_reconnect_soon (#117945)

autinerd committed:
Bump ruff to 0.4.5 (#117958)

jpbede committed:
Allow to reconfigure integrations with `single_config_entry` set (#117939)

Galorhallen committed:
Better handling of EADDRINUSE for Govee light (#117943)

kaareseras committed:
Add Azure data explorer (#68992)

Co-authored-by: Robert Resch <robert@resch.dev>

farmio committed:
Fix run-in-env script for not running in venv (#117961)

bdraco committed:
Simplify async_track_time_interval implementation (#117956)

joostlek committed:
Bump airgradient to 0.4.1 (#117963)

dontinelli committed:
Improve fyta tests (#117661)

* Add test for init

* update tests

* split common.py into const.py and __init__.py

* Update tests/components/fyta/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* add autospec, tidy up

* adjust len-test

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

epenet committed:
Cleanup pylint ignore (#117964)

epenet committed:
Add snapshot tests to moehlenhoff_alpha2 (#117967)

* Add tests to moehlenhoff_alpha2

* Adjust coverage

* Adjust coverage

* Adjust coverage

* Adjust patch

* Adjust

Bre77 committed:
Add switch platform to Teslemetry (#117482)

* Add switch platform

* Add tests

* Add test

* Fixes

* ruff

* Rename to storm watch

* Remove valet

* Apply suggestions from code review

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* ruff

* Review feedback

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

epenet committed:
Use SnapshotAssertion in rainmachine diagnostic tests (#117979)

* Use SnapshotAssertion in rainmachine diagnostic tests

* Force entry_id

* Adjust

* Fix incorrect fixtures

* Adjust

epenet committed:
Add snapshot platform tests to rainmachine (#117978)

* Add snapshot tests to rainmachine

* Add sensor and switch tests

epenet committed:
Move rainmachine coordinator to separate module (#117983)

* Move rainmachine coordinator to separate module

* Coverage

bdraco committed:
Fix turbojpeg init doing blocking I/O in the event loop (#117971)

* Fix turbojpeg init doing blocking I/O in the event loop

fixes
```
Detected blocking call to open inside the event loop by integration camera at homeassistant/components/camera/img_util.py, line 100: TurboJPEGSingleton.__instance = TurboJPEG() (offender: /usr/local/lib/python3.12/ctypes/util.py, line 276: with open(filepath, rb) as fh:), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+camera%22
```

* Update homeassistant/components/camera/img_util.py

* Fix turbojpeg init doing blocking I/O in the event loop

fixes
```
Detected blocking call to open inside the event loop by integration camera at homeassistant/components/camera/img_util.py, line 100: TurboJPEGSingleton.__instance = TurboJPEG() (offender: /usr/local/lib/python3.12/ctypes/util.py, line 276: with open(filepath, rb) as fh:), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+camera%22
```

* already suppressed and logged

bdraco committed:
Cache serialize of manifest for loaded integrations (#117965)

* Cache serialize of manifest for loaded integrations

The manifest/list and manifest/get websocket apis
are called frequently when moving around in the UI.
Since the manifest does not change we can make
the the serialized version a cached property

* reduce

* reduce

agrauballe committed:
Deconz - Added trigger support for Aqara WB-R02D mini switch (#117917)

Added support for Aqara WB-R02D mini switch

Co-authored-by: agr <agr@kamstrup.com>

allenporter committed:
Bump to google-nest-sdm to 4.0.4 (#117982)

thecode committed:
Enable Switcher assume buttons for all devices (#117993)

bdraco committed:
Bump habluetooth to 3.1.1 (#117992)

epenet committed:
Add snapshot tests to omnilogic (#117986)

epenet committed:
Move moehlenhoff_alpha2 coordinator to separate module (#117970)

thecode committed:
Remove Switcher YAML import support (#117994)

balloob committed:
Google gen updates (#117893)

* Add a recommended model for Google Gen AI

* Add recommended settings to Google Gen AI

* Revert no API msg

* Use correct default settings

* Make sure options are cleared when using recommended

* Update snapshots

* address comments

bdraco committed:
Update pySwitchbot to 0.46.0 to fix lock key retrieval (#118005)

* Update pySwitchbot to 0.46.0 to fix lock key retrieval

needs https://github.com/Danielhiversen/pySwitchbot/pull/236

* bump

* fixes

cdce8p committed:
Update codespell to 2.3.0 (#118001)

Bre77 committed:
Add lock platform to Teslemetry (#117344)

* Add lock platform

* Tests and fixes

* Fix json

* Review Feedback

* Apply suggestions from code review

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* wording

* Fix rebase

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

Thomas55555 committed:
Remove unused snapshots [a-f] (#117999)

cdce8p committed:
Fix vallow test fixtures (#118003)

epenet committed:
Move omnilogic coordinator to separate module (#118014)

Bre77 committed:
Add binary sensor platform to Teslemetry (#117230)

* Add binary sensor platform

* Add tests

* Cleanup

* Add refresh test

* Fix runtime_data after rebase

* Remove streaming strings

* test error

* updated_once

* fix updated_once

* assert_entities_alt

* Update homeassistant/components/teslemetry/binary_sensor.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thecode committed:
Switcher - use single_config_entry and register_discovery_flow in con… (#118000)

Bre77 committed:
Add models to energy sites in Teslemetry (#117419)

* Add models to energy sites and test devices

* Fix device testing

* Revert VIN

* Fix snapshot

* Fix snapshot

* fix snap

* Sort list

epenet committed:
Add snapshot tests to plaato (#118017)

kingy444 committed:
Daikin Aircon - Add strings and debug (#116674)

frenck committed:
Add sequence action for automations & scripts (#117690)

Co-authored-by: Robert Resch <robert@resch.dev>

epenet committed:
Move plaato coordinator to separate module (#118019)

epenet committed:
Use snapshot in renault diagnostics tests (#118021)

freekode committed:
Migrate OpenWeaterMap to new library (support API 3.0) (#116870)

Co-authored-by: J. Nick Koston <nick@koston.org>

emontnemery committed:
Improve async_get_issue_tracker for custom integrations (#118016)

epenet committed:
Add logging to SamsungTV turn-on (#117962)

* Add logging to SamsungTV turn-on

* Apply suggestions from code review

Co-authored-by: J. Nick Koston <nick@koston.org>

---------

Co-authored-by: J. Nick Koston <nick@koston.org>

cdce8p committed:
Update typing-extensions to 4.12.0 (#118020)

bdraco committed:
Avoid calling split_entity_id in event add/remove filters (#118015)

cdce8p committed:
Improve callable annotations (#118024)

cdce8p committed:
Fix partial typing (#118022)

cdce8p committed:
Update codespell ignore list (#118018)

fredrike committed:
Fix issue with device_class.capitalize() in point (#117969)

Ulfmerbold2000 committed:
Add missing Ecovacs life spans (#117134)

Co-authored-by: Robert Resch <robert@resch.dev>

epenet committed:
Add snapshot tests to nuki (#117973)

bdraco committed:
Bump pySwitchbot to 0.46.1 (#118025)

epenet committed:
Add error message to snapshot_platform helper (#117974)

mkmer committed:
Catch client connection error in Honeywell (#117502)

Co-authored-by: Robert Resch <robert@resch.dev>

PeteRager committed:
Fix Sonos album artwork performance (#116391)

autinerd committed:
Convert namedtuple to NamedTuple for smartthings (#115395)

Thomas55555 committed:
Refactor Husqvarna Automower (#117938)

epenet committed:
Move nuki coordinator to separate module (#117975)

epenet committed:
Move TibberDataCoordinator to separate module (#118027)

jbouwh committed:
Create bound callback_message_received method for handling mqtt callbacks (#117951)

* Create bound callback_message_received method for handling mqtt callbacks

* refactor a bit

* fix ruff

* reduce overhead

* cleanup

* cleanup

* Revert changes alarm_control_panel

* Add sensor and binary sensor

* use same pattern for MqttAttributes/MqttAvailability

* remove useless function since we did not need to add to it

* code cleanup

* collapse

---------

Co-authored-by: J. Nick Koston <nick@koston.org>

Bre77 committed:
Add Cover platform to Teslemetry (#117340)

* Add cover

* Test coverage

* Json lint

* Apply suggestions from code review

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* Update tests

* Fix features

* Update snapshot from fixture

* Apply suggestions from code review

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

esciara committed:
Remove duplicate tests in generic_thermostat (#105622)

Tests using `setup_comp_4` and `setup_comp_6` have been replaced by a parameterized tests in #105643. Tests using `setup_comp_5` are therefore still duplicates and are removed.

jbouwh committed:
Improve error logging on invalid MQTT entity state (#118006)

* Improve error logging on invalid MQTT entity state

* Explain not hanlding TpeError and ValueError

* Move length check closer to source

* use _LOGGER.exception

PhilRW committed:
Add APRS object tracking (#113080)

* Add APRS object tracking

Closes issue #111731

* Fix unit test

---------

Co-authored-by: Erik Montnemery <erik@montnemery.com>

FWeinb committed:
Make co/co2 threshold configurable via entity_config (#112978)

* make co/co2 threshold configurable via entity_config

* Split threshold into co/co2_threshold configuration

marcelveldt committed:
Extend the blocklist for Matter transitions with more models (#118038)

yuvalabou committed:
Add config flow to Jewish Calendar (#84464)

* Initial commit

* add basic tests (will probably fail)

* Set basic UID for now

* Various improvements

* use new naming convention?

* bit by bit, still not working tho

* Add tz selection

* Remove failing tests

* update unique_id

* add the tests again

* revert to previous binary_sensor test

* remove translations

* apply suggestions

* remove const.py

* Address review

* revert changes

* Initial fixes for tests

* Initial commit

* add basic tests (will probably fail)

* Set basic UID for now

* Various improvements

* use new naming convention?

* bit by bit, still not working tho

* Add tz selection

* Remove failing tests

* update unique_id

* add the tests again

* revert to previous binary_sensor test

* remove translations

* apply suggestions

* remove const.py

* Address review

* revert changes

* Fix bad merges in rebase

* Get tests to run again

* Fixes due to fails in ruff/pylint

* Fix binary sensor tests

* Fix config flow tests

* Fix sensor tests

* Apply review

* Adjust candle lights

* Apply suggestion

* revert unrelated change

* Address some of the comments

* We should only allow a single jewish calendar config entry

* Make data schema easier to read

* Add test and confirm only single entry is allowed

* Move OPTIONS_SCHEMA to top of file

* Add options test

* Simplify import tests

* Test import end2end

* Use a single async_forward_entry_setups statement

* Revert schema updates for YAML schema

* Remove unneeded brackets

* Remove CONF_NAME from config_flow

* Assign hass.data[DOMAIN][config_entry.entry_id] to a local variable before creating the sensors

* Data doesn't have a name remove slugifying of it

* Test that the entry has been created correctly

* Simplify setup_entry

* Use suggested values helper and flatten location dictionary

* Remove the string for name exists as this error doesn't exist

* Remove name from config entry

* Remove _attr_has_entity_name - will be added in a subsequent PR

* Don't override entity id's - we'll fixup the naming later

* Make location optional - will by default revert to the user's home location

* Update homeassistant/components/jewish_calendar/strings.json

Co-authored-by: Erik Montnemery <erik@montnemery.com>

* No need for local lat/long variable

* Return name attribute, will deal with it in another PR

* Revert unique_id changes, will deal with this in a subsequent PR

* Add time zone data description

* Don't break the YAML config until the user has removed it.

* Cleanup initial config flow test

---------

Co-authored-by: Tsvi Mostovicz <ttmost@gmail.com>
Co-authored-by: Erik Montnemery <erik@montnemery.com>

bdraco committed:
Add json cache to lovelace config (#117843)

cdce8p committed:
Improve annotation styling (#118032)

epenet committed:
Use issue_registry fixture in core tests (#118042)

brg468 committed:
Update Rachio hose timer battery sensor (#118045)

jbouwh committed:
Rename MQTT mixin classes (#118039)

edenhaus committed:
Remove strict connection (#117933)

epenet committed:
Use issue_registry fixture in component tests (#118041)

dontinelli committed:
Change ZoneInfo to async_get_time_zone in fyta (#117996)

bdraco committed:
Add types to event tracker data (#118010)

* Add types to event tracker data

* fixes

* do not test event internals in other tests

* fixes

* Update homeassistant/helpers/event.py

* cleanup

* cleanup

epenet committed:
Exclude gold and platinum integrations from .coveragerc (#117563)

synesthesiam committed:
Fix intent helper test (#118053)

Fix test

synesthesiam committed:
Require registered device id for all timer intents (#117946)

* Require device id when registering timer handlers

* Require device id for timer intents

* Raise errors for unregistered device ids

* Add callback

* Add types for callback to __all__

* Clean up

* More clean up

balloob committed:
Remove unneeded asserts (#118056)

* Remove unneeded asserts

* No need to guard changing a timer that is owned by a disconnected device

balloob committed:
Update OpenAI defaults (#118059)

* Update OpenAI defaults

* Update max temperature

balloob committed:
Allow llm API to render dynamic template prompt (#118055)

* Allow llm API to render dynamic template prompt

* Make rendering api prompt async so it can become a RAG

* Fix test

NECH2004 committed:
Add tests to Zeversolar integration (#117928)

jbouwh committed:
Refactor mqtt callbacks for alarm_control_panel (#118037)

jbouwh committed:
Refactor mqtt callbacks for cover (#118044)

bachya committed:
Add activity type to appropriate RainMachine switches (#117875)

bdraco committed:
Significantly speed up single use callback dispatchers (#117934)

jbouwh committed:
Refactor mqtt callbacks for climate and water_heater (#118040)

* Refactor mqtt callbacks for climate and water_heater

* Reduce callbacks

jbouwh committed:
Move recorder chunk utils to shared collection utils (#118065)

jbouwh committed:
Consequently ignore empty MQTT state payloads and set state to `unknown` on "None" payload (#117813)

* Consequently ignore empty MQTT state payloads and set state to `unknown` on "None" payload

* Do not change preset mode behavior

* Add device tracker ignoring empty state

* Ignore empty state for lock

* Resolve merge errors

jbouwh committed:
Split mqtt subscribe and unsubscribe calls to smaller chunks (#118035)

bdraco committed:
Avoid generating matchers that will never be used in MQTT (#118068)

bdraco committed:
Avoid catch_log_exception overhead in MQTT for simple callbacks (#118036)

bdraco committed:
Reduce duplicate code to handle mqtt message replies (#118067)

bdraco committed:
Use defaultdict instead of setdefault in mqtt client (#118070)

jbouwh committed:
Fix lingering mqtt test (#118072)

tronikos committed:
Google Generative AI: add timeout to ensure we don't block HA startup (#118066)

* timeout

* fix

* tests

tronikos committed:
Add Google Generative AI safety settings (#117679)

* Add safety settings

* snapshot-update

* DROPDOWN

* fix test

* rename const

* Update const.py

* Update strings.json

Shulyaka committed:
Add user name and location to the LLM assist prompt (#118071)

Add user name and location to the llm assist prompt

balloob committed:
Add device info to Google (#118074)

bdraco committed:
Speed up is_template_string by avoiding regex engine (#118076)

bdraco committed:
Remove OrderedDict from entity_values as dict guarantees order on newer cpython (#118081)

bdraco committed:
Speed up async_render_with_possible_json_value (#118080)

bdraco committed:
Avoid template context manager overhead when template is already compiled (#118079)

bdraco committed:
Remove unused code in template helper (#118075)

bdraco committed:
Convert remaining mqtt attrs classes to dataclasses (#118073)

balloob committed:
Add device info to OpenAI (#118077)

balloob committed:
Fix default Google AI prompt on initial setup (#118078)

balloob committed:
Add no-API LLM prompt back to Google (#118082)

* Add no-API LLM prompt back

* Use string join

balloob committed:
Add recommended model options to OpenAI (#118083)

* Add recommended options to OpenAI

* Use string join

balloob committed:
Update Google safety defaults to match Google (#118084)

balloob committed:
Flip prompts to put user prompt on top (#118085)

cereal2nd committed:
Bump velbusaio to 2024.5.1 (#118091)

bdraco committed:
Speed up removing MQTT subscriptions (#118088)

mib1185 committed:
Store runtime data inside the config entry in Spotify (#117037)

mib1185 committed:
Store runtime data inside the config entry in Forecast Solar (#117033)

elupus committed:
Adjust title of integration sensor (#116954)

allistermaguire committed:
Bump pyenvisalink version to 4.7 (#118086)

bdraco committed:
Simplify mqtt discovery cooldown calculation (#118095)

bdraco committed:
Migrate firmata to use async_unload_platforms (#118098)

bdraco committed:
Migrate vera to use async_unload_platforms (#118099)

bdraco committed:
Migrate zha to use async_unload_platforms (#118100)

bdraco committed:
Migrate zwave_js to use async_unload_platforms (#118101)

bdraco committed:
Switch mqtt to use async_unload_platforms (#118097)

thecode committed:
Store Switcher runtime data in config entry (#118054)

michaelkkehoe commented:
Polite bump

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    coordinator = SwidgetDataUpdateCoordinator(hass, device)
+    try:
+        await coordinator.async_config_entry_first_refresh()
+    except Exception as ex:
+        raise ConfigEntryNotReady(
+            f"Unable to to perform initial refresh: {entry.data[CONF_HOST]}"
+        ) from ex
```
catching general `Exception` is only allowed in the config_flow. Also I'd move that inside the coordinator then, since you're already catching there.

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    """Unload a config entry."""
+    hass_data: dict[str, Any] = hass.data[DOMAIN]
+    device: SwidgetDevice = hass_data[entry.entry_id].device
+    await device.stop()
```
can this throw?

michaelkkehoe replied:
No

zweckj commented on homeassistant/components/swidget/__init__.py:
```diff
@@ -0,0 +1,71 @@
+        raise ConfigEntryNotReady(
+            f"Unable to to perform initial refresh: {entry.data[CONF_HOST]}"
+        ) from ex
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
```suggestion
    entry.runtime_data = coordinator
```
is the new way and define `type SwidgetConfigEntry = ConfigEntry[SwidgetDataUpdateCoordinator]` and replace all ConfigEntry with `SwidgetConfigEntry`

michaelkkehoe replied:
Updated it. That wasn't fun.

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,92 @@
+        await device.stop()
+    except SwidgetException as exc:
+        raise CannotConnect from exc
+    friendly_name: str = device.friendly_name
+    unique_id: str = f"{device.id}_{device.mac_address}"
+    return {"title": friendly_name, "unique_id": unique_id}
```
```suggestion
    return {"title": device.friendly_name, "unique_id": f"{device.id}_{device.mac_address}"}
```
since you're not reusing those, it's a bit more readable imo

michaelkkehoe replied:
Fixed

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,92 @@
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+                unique_id = info["unique_id"]
```
move this out of the `try`

michaelkkehoe replied:
I'm not sure I follow on this one. Why wouldn't I want the `try` here?

zweckj commented on homeassistant/components/swidget/config_flow.py:
```diff
@@ -0,0 +1,92 @@
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
+
+
+class InvalidHost(HomeAssistantError):
+    """Error to indicate the host is not valid."""
+
+
+class TimeoutConnect(HomeAssistantError):
+    """Error to indicate the connection attempt is timed out."""
```
```suggestion
```
appear to be unused

zweckj commented on homeassistant/components/swidget/coordinator.py:
```diff
@@ -0,0 +1,54 @@
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+_LOGGER = logging.getLogger(__name__)
+REQUEST_REFRESH_DELAY = 1.0
```
```suggestion
```
appears unused

zweckj commented on homeassistant/components/swidget/coordinator.py:
```diff
@@ -0,0 +1,54 @@
+            ),
+        )
+
+    async def async_initialize(self) -> Any:
+        """Initialize a callback for any websocket events received from the device."""
+        return self.device.add_event_callback(self.websocket_event_callback)
```
Iit's a bit weird that you're doing some initialization inside the init, but then some here. I'd move it in one place

zweckj commented on homeassistant/components/swidget/entity.py:
```diff
@@ -0,0 +1,59 @@
+
+    async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        await func(self, *args, **kwargs)
+        # await asyncio.sleep(1)
```
remove commented out code

michaelkkehoe replied:
Fixed. Sorry

zweckj commented on homeassistant/components/swidget/light.py:
```diff
@@ -0,0 +1,87 @@
+        # Fallback to adjusting brightness or turning the bulb on
+        if brightness is not None:
+            await self.device.set_brightness(brightness)
+            # await self.coordinator._async_update_data()
```
remove commented out code in this file

michaelkkehoe replied:
Fixed. Sorry

zweckj commented on homeassistant/components/swidget/entity.py:
```diff
@@ -0,0 +1,59 @@
+_P = ParamSpec("_P")
+
+
+def async_refresh_after(
+    func: Callable[Concatenate[_T, _P], Awaitable[None]],
+) -> Callable[Concatenate[_T, _P], Coroutine[Any, Any, None]]:
+    """Define a wrapper to refresh after."""
+
+    async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        await func(self, *args, **kwargs)
+        # await asyncio.sleep(1)
```
appears unused, also the generics (typevars) appear to be unused, so please remove them until you actually use them

zweckj commented on homeassistant/components/swidget/light.py:
```diff
@@ -0,0 +1,87 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+BRIGHTNESS = "brightness"
+BRIGHTNESS_SCALE = (0, 100)
+VAL = vol.Range(min=0, max=100)
+SWIDGET_SET_BRIGHTNESS_SCHEMA = cv.make_entity_service_schema({BRIGHTNESS: VAL})
```
```suggestion
BRIGHTNESS_SCALE = (0, 100)
```
those appear unused

michaelkkehoe replied:
Fixed

zweckj commented on tests/components/swidget/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        await hass.async_block_till_done()
+    assert result2["type"] is FlowResultType.FORM
+    assert result2["errors"] == {"base": "unknown"}
+    assert len(mock_setup_entry.mock_calls) == 0
```
for non-terminal states, please fix the error and finish the test successfully to show we can recover from the error
